Shell/AST.h

```diff
@@ -0,0 +1,730 @@
+    virtual HitTestResult hit_test_position(size_t offset)
+    {
+        if (m_position.in_range(offset))
+            return { this, nullptr };
+        return { nullptr, nullptr };
+    }
+    virtual String class_name() const { return "Node"; }
+    Node(Position);
+    virtual ~Node();
+
+    virtual bool is_bareword() const { return false; }
+    virtual bool is_command() const { return false; }
+    virtual bool is_execute() const { return false; }
+    virtual bool is_glob() const { return false; }
+    virtual bool is_variable_decl() const { return false; }
+
+    virtual bool is_list() const { return false; }
+
+    const Position& position() const { return m_position; }
+
+protected:
+    Position m_position;
+};
+class And : public Node {
+public:
+    virtual void dump(int level) override;
+    virtual RefPtr<Value> run(TheExecutionInputType) override;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) override;
+    virtual HitTestResult hit_test_position(size_t) override;
+    virtual String class_name() const override { return "And"; }
+    And(Position, RefPtr<Node>, RefPtr<Node>);
+    virtual ~And();
+
+private:
+    RefPtr<Node> m_left;
+    RefPtr<Node> m_right;
+};
+
+class PrependToList : public Node {
+public:
+    virtual void dump(int level) override;
+    virtual RefPtr<Value> run(TheExecutionInputType) override;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) override;
+    virtual HitTestResult hit_test_position(size_t) override;
+    virtual String class_name() const override { return "PrependToList"; }
+    virtual bool is_list() const override { return true; }
+    PrependToList(Position, RefPtr<Node>, RefPtr<Node>);
+    virtual ~PrependToList();
+
+private:
+    RefPtr<Node> m_element;
+    RefPtr<Node> m_list;
+};
+
+class Background : public Node {
+public:
+    virtual void dump(int level) override;
+    virtual RefPtr<Value> run(TheExecutionInputType) override;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) override;
+    virtual HitTestResult hit_test_position(size_t) override;
+    virtual String class_name() const override { return "Background"; }
+    Background(Position, RefPtr<Node>);
+    virtual ~Background();
+
+private:
+    RefPtr<Node> m_command;
+};
+
+class BarewordLiteral : public Node {
+public:
+    virtual void dump(int level) override;
+    virtual RefPtr<Value> run(TheExecutionInputType) override;
+    virtual void highlight_in_editor(Line::Editor&, RefPtr<Shell>, HighlightMetadata = {}) override;
+    virtual String class_name() const override { return "BarewordLiteral"; }
+    BarewordLiteral(Position, String);
+    virtual ~BarewordLiteral();
+    virtual bool is_bareword() const override { return true; }
+    const String& text() const { return m_text; }
+
+private:
+    String m_text;
+};
+
+class CastToCommand : public Node {
```

Comment:
Are cast-to-command and cast-to-list distinct things? I thought they wouldn't be...

Reply:
of course, a command is a list until used in command context.